### PR TITLE
dns_filter: Generate responses to queries

### PR DESCRIPTION
**Pull request by abaptiste**
Internal PR: Adding response generation to DNS Filter

### CODE REVIEW

File: api/envoy/data/dns/v3/dns_table.proto [hunk 1]
```diff
@@ -59,27 +56,25 @@ message DnsTable {
     option (udpa.annotations.versioning).previous_message_type =
         "envoy.data.dns.v2alpha.DnsTable.DnsVirtualDomain";
 
-    // The domain name for which Envoy will respond to query requests
+    // A domain name for which Envoy will respond to query requests
     string name = 1 [(validate.rules).string = {min_len: 2 well_known_regex: HTTP_HEADER_NAME}];
```
Comment (lavignes) [inline api/envoy/data/dns/v3/dns_table.proto:60]: thats an odd regex for a DNS name

File: source/extensions/filters/udp/dns_filter/dns_filter.cc [hunk 2]
```diff
@@ -76,21 +109,149 @@ void DnsFilter::onData(Network::UdpRecvData& client_request) {
...
+  const auto& known_suffixes = config_->knownSuffixes();
+
+  // If we don't have a list of whitelisted domain suffixes, we will resolve the name with an
+  // external DNS server
+  if (known_suffixes.empty()) {
+    ENVOY_LOG(debug, "Known domains list is empty");
+    return false;
+  }
+
+  // TODO: Use a trie to find a match instead of iterating through the list
```
Comment (lavignes) [inline source/extensions/filters/udp/dns_filter/dns_filter.cc:195]: nit: Should tag TODOs `TODO(abaptiste): `

File: source/extensions/filters/udp/dns_filter/dns_filter.cc [hunk 3]
```diff
@@ -3,55 +3,59 @@
...
-        const auto& address_list = virtual_domain.endpoint().address_list().address();
-        addrs.reserve(address_list.size());
-        // This will throw an exception if the configured_address string is malformed
-        for (const auto& configured_address : address_list) {
-          const auto ipaddr =
-              Network::Utility::parseInternetAddress(configured_address, 0 /* port */);
-          addrs.push_back(ipaddr);
-        }
+  envoy::data::dns::v3::DnsTable dns_table;
+  bool result = loadServerConfig(server_config, dns_table);
```
Comment (lavignes) [inline source/extensions/filters/udp/dns_filter/dns_filter.cc:28]: question: How does this failure get signaled to the control-plane? Like if a user supplies bad config from an xDS endpoint. Dont we need to throw an exception or something? Not sure the correct API for it...

Comment (nealp9084) [reply source/extensions/filters/udp/dns_filter/dns_filter.cc:28]: I would recommend asking Marco. He implemented something like that for the lambda filter, where Envoy would just NACK a bad configuration.

File: source/extensions/filters/udp/dns_filter/dns_filter.cc (hunk 2 above)
Comment (abaptiste) [reply source/extensions/filters/udp/dns_filter/dns_filter.cc:195]: Sure.  I'll update

File: api/envoy/data/dns/v3/dns_table.proto (hunk 1 above)
Comment (abaptiste) [reply api/envoy/data/dns/v3/dns_table.proto:60]: https://github.com/envoyproxy/protoc-gen-validate/blob/ab56c3dd1cf9b516b62c5087e1ec1471bd63631e/validate/validator.py#L19

Looks reasonable to me

File: source/extensions/filters/udp/dns_filter/dns_filter.cc (hunk 3 above)
Comment (abaptiste) [reply source/extensions/filters/udp/dns_filter/dns_filter.cc:28]: I'll take a look and see what I can find.

Comment (abaptiste) [reply source/extensions/filters/udp/dns_filter/dns_filter.cc:28]: The 'convention' seems to be that we throw an EnvoyException.   The code I used to load the data source was mirrored from elsewhere in Envoy, complete with the exception catching.

The filter does work even if the data source isn't read.  There are messages in the logs that will indicate a failure.

File: api/envoy/data/dns/v3/dns_table.proto (hunk 1 above)
Comment (lavignes) [reply api/envoy/data/dns/v3/dns_table.proto:60]: Well I guess odd regex name :)

File: source/extensions/filters/udp/dns_filter/dns_filter.cc (hunk 3 above)
Comment (lavignes) [reply source/extensions/filters/udp/dns_filter/dns_filter.cc:28]: > The filter does work even if the data source isn't read

Ah I see.

File: api/envoy/data/dns/v3/dns_table.proto (hunk 1 above)
Comment (abaptiste) [reply api/envoy/data/dns/v3/dns_table.proto:60]: That I can agree on

**Pull request closed without merging**